---
title: "thesis_mbds"
author: "Sebastian Gritsch"
date: "2023-11-14"
output:
  html_document: default
  pdf_document: default
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r initialize_project}
source(here::here("InitializeProject.R"))
library(tidyverse); packageVersion("tidyverse")
library(magrittr); packageVersion("magrittr")
```
### Generate manifest file for FragPipe
```{r generate fragpipe manifest}
# Get list of mzML files with full path names
ms_data_dir <- normalizePath("data/ms_data/mzml")
ms_files <- list.files(ms_data_dir, pattern = "mzML", full.names = TRUE)

# create data frame for fragpipe manifest
# extract experiments (condition_time) from file names 
fp_manifest <- data.frame(file = ms_files, stringsAsFactors = FALSE) %>%
  mutate(experiment = str_extract(file, "N_[A-Z]_t[0-5]"),
         # extract bioreplicates from file names
         bioreplicate = gsub(".mzML", "", str_extract(file, "[1-3].mzML$")),
         # add column with data type "DDA"
         data_type = "DDA") %>%
  # select rows where experiment is not NA
  filter(!is.na(experiment)) %>%
  set_colnames(c("file", "experiment", "bioreplicate", "data_type"))

# export tab separated file with the name "fragpipe-files.fp-manifest"
write.table(fp_manifest, file = here::here("data/fragpipe-files.fp-manifest"),
            sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)
```


### Download contaminants database

Download Cambridge Centre for Proteomics cRAP FASTA database (CCP cRAP)
```{r download cRAP database}
#remotes::install_github("CambridgeCentreForProteomics/camprotR", dependencies = TRUE)
#library(camprotR); package.version('camprotR')
#download_ccp_crap(paste0("data/prot_db/raw_db_files/",check_uniprot_release(), "_CCP_cRAP.fasta"), is_crap = TRUE, overwrite = FALSE, verbose = TRUE)
```
> Downloading from UniProtKB release: 2023_05

This databse will be concatinated with a protein database downloaded from Uniprot. Decoys will be added to the combined database using Philosopher and the resulting database will be used for the subsequent analysis. Proteins were identified using the FragPipe suite with LFQ. The output file "combined_protein.tsv" is used for the subsequent differential expression analysis.
```{r}
datadir <- "results/00_fragpipe/240210"

```


### Anlysis in LIMMA

#### Load data into an ExpressionSet object
```{r load_data}
library(Biobase); packageVersion("Biobase")


# Import fragpipe results
data <- rio::import(here::here(datadir, "combined_protein.tsv"))

# filter out contaminating proteins from cRAP database
data <- data %>% filter(!grepl("cRAP[0-9]{3}", Protein))

# Extract protein annotation from fragpipe output 'data'
feature_annotation <- data %>%
  dplyr::select('Protein',
                'Protein ID',
                'Entry Name',
                'Gene',
                'Protein Length',
                'Organism',
                'Protein Existence',
                'Description',
                'Protein Probability',
                'Top Peptide Probability') %>%
   mutate(rowname = data$'Protein ID') %>%
   column_to_rownames()

# Extract protein intensities per sample from fragpipe output 'data'
assayData <- data %>%
  dplyr::select('Protein ID', ends_with("Intensity"), -contains("MaxLFQ")) %>%
  column_to_rownames('Protein ID')

# Remove 'Intensity' from column names
colnames(assayData) <- colnames(assayData) %>%
  str_replace(" Intensity", "")

# log2-transform intensities
assayData[assayData == 0] <- 1
assayData <- assayData %>% 
  log2

# Import sample annotation from fragpipe annotation file
sample_annotation <- rio::import(here::here(datadir, "experiment_annotation.tsv"))

# Rename condition. this will change once the proper "experiment_annotation.tsv" is used
sample_annotation <- sample_annotation %>%
  dplyr::select(sample, condition) %>%
  set_colnames(c("SAMPLEID", "CONDITION")) %>%
  # extract information from sample names using regular expressions
  mutate(CONDITION = gsub("[NS]_", "", str_extract(SAMPLEID, "[NS]_[A-Z]")),
         REPLICATE = gsub("_", "", str_extract(SAMPLEID, "_[1-3]$")),
         TIME = str_extract(SAMPLEID, "t[0-5]"),
         FACTOR = str_extract(SAMPLEID, "[A-Z]_t[0-5]")) %>%
  mutate(rowname = SAMPLEID) %>%
  column_to_rownames()

# Construct eset from 'assayData'
eset <- ExpressionSet(as.matrix(assayData))

# Add metadata ('sampleNames') and featuredata 'featureNames' to eset
pData(eset) <- sample_annotation[sampleNames(eset),]
fData(eset) <- feature_annotation[featureNames(eset),]

# Check validity of eset object
validObject(eset)

# Save eset
eset %>%
  rio::export(here::here(datadir, "eset.RDS"))
```
#### Linear model with limma only considering the condition

This analysis does not consider time, which is likely to have an effect on protein expression. This is a preliminary analysis to identify proteins that are differentially expressed between conditions. A more comprehensive analysis will be performed later, considering both condition and time.


```{r linear model with limma}
library(limma); packageVersion("limma")

# Import eset
eset <- rio::import(here::here(datadir, "eset.RDS"))

# Convert 'CONDITION' to factor
eset$CONDITION <- as.factor(eset$CONDITION)
eset$TIME <- as.factor(eset$TIME)

# Construct design matrix
design <- model.matrix(~0 + CONDITION, data = eset)
colnames(design) <- levels(eset$CONDITION)

# Define conditions for contrast matrix
contrasts <- list()
# loop through conditions except the control "A" to define contrasts
for (condition in levels(eset$CONDITION)[-1]) {
  contrast_name <- paste(condition, "VS", "A", sep = "_")
  contrast <- paste(condition, "A", sep = " - ")
  contrasts[[contrast_name]] <- contrast
}

# Construct constrast matrix
contrast_matrix <- makeContrasts(contrasts = contrasts, levels = design)
colnames(contrast_matrix) <- names(contrasts)

# Define linear model
fit <- eset %>%
  # linear regression of every protein against designmatrix as covariable
  lmFit(design) %>%
  # include constrast matrix
  contrasts.fit(contrast_matrix) %>%
  # Empirical Bayes Statistics for Differential Expression
  eBayes

# Extract levels of contrast matrix
coefs <- names(contrasts) %>%
  set_names(.,.)

# Generate table of top-ranking features from the fitted model for each level
limma_results <- coefs %>%
  purrr::map(function(coef, fit){
    topTable(fit, 
             coef = coef, 
             number = Inf)
  },
  fit = fit)

# Save results table of each level on separate worksheets in a Excel file
limma_results %>%
  rio::export(here::here(paste(resultsdir,"01_limma", "limma_results.xlsx", sep="/"))) 
```

```{r Explore Limma results}
# Import Limma results
#limma_results <- rio::import_list(here::here(resultsdir, "limma_results.xlsx"))

# Count significantly deregulated proteins per contrast
lapply(limma_results, function(df) {
  df$Protein.ID[df$adj.P.Val <= 0.1 & abs(df$logFC) > 2] %>% length()})
# lower significance threshold to 0.5
lapply(limma_results, function(df) {
  df$Protein.ID[df$adj.P.Val <= 0.5 & abs(df$logFC) > 2] %>% length()})
# lower significance threshold to 1
lapply(limma_results, function(df) {
  df$Protein.ID[df$adj.P.Val <= 1 & abs(df$logFC) > 2] %>% length()})
```

```{r}
top_proteins <- topTable(fit, number = 30)
top_proteins

# make heatmap from top table
top_proteins %>%
  # filter columns for heatmap
  select(contains("_VS_")) %>%
  # convert to matrix
  as.matrix() %>%
  heatmap(.,
          Colv = NA, # no column clustering
          Rowv = NA, # no row clustering
          labRow = top_proteins$Description, # row labels
          cexCol = 1, cexRow = 0.5, # axis label size
          main = "Most significantly deregulated proteins")
```
#### Linear model with condition and time
```{r linear model with limma: condition and time}

# Convert 'CONDITION' to factor
eset$CONDITION <- as.factor(eset$CONDITION)
eset$TIME <- as.factor(eset$TIME)
eset$FACTOR <- as.factor(eset$FACTOR)

# factor condition_time 
# design (~0 + factor)
# contrasts
# condA_t2 - condA_t1, ...
# condB_t2 - condB_t1, ...
# condA_t1 - condB_t1

# Construct design matrix
design2 <- model.matrix(~0 + FACTOR, data = eset)
# set names for the design matrix columns
colnames(design2) <- levels(eset$FACTOR)


# Initiate contrast list
contrasts <- list()

# For each condition make contrast of time points: A_t0-A_t1, A_t3-A_t1, ...
# Loop through conditions
for (condition in levels(eset$CONDITION)){
  # initiate prev_factor as first factor in the condition
  prev_factor <- grep(condition, levels(eset$FACTOR), value = TRUE)[1]
  # loop through remaining factors of the condition and build contrasts
  for (factor in grep(condition, levels(eset$FACTOR), value = TRUE)[-1]){
    # Create contrast name and contrast
    contrast_name <- paste(prev_factor, "vs", factor, sep = "_")
    contrast <- paste(prev_factor, factor, sep = " - ")
    # Store contrast in the list
    contrasts[[contrast_name]] <- contrast
    # Update prev_factor
    prev_factor <- factor
  }
}

# For each time point make contrasts of factors: A_t1-B_t1, A_t1-C_t1, ...
# Loop through time points
for (t in levels(eset$TIME)){
  # Iterate through each combination of factors
  for(factor in grep(t, levels(eset$FACTOR), value = TRUE)){
    for (factor2 in grep(t, levels(eset$FACTOR), value = TRUE)){
      # Check if factors are not identical
      if (!identical(factor, factor2)){
        # Create contrast name and contrast
        contrast_name <- paste(substr(factor,1, 1), "vs", factor2, sep = "_")
        contrast <- paste(factor, factor2, sep = " - ")
        # Store contrast in the list
        contrasts[[contrast_name]] <- contrast
      }
    }
  }
}

# For each time point create contrasts from unique combinations of factors
# # Loop through time points
# for (t in levels(eset$TIME)) {
#   # Subset factors for the current time point
#   factors <- grep(t, levels(eset$FACTOR), value = TRUE)
#   
#   # Generate all unique combinations of factors
#   factor_combinations <- combn(factors, 2, simplify = FALSE)
#   
#   # Iterate over each combination
#   for (combination in factor_combinations) {
#     factor1 <- combination[1]
#     factor2 <- combination[2]
#     
#     # Create contrast name and contrast
#     contrast_name <- paste(substr(factor1, 1, 1), "VS", factor2, sep = "_")
#     contrast <- paste(factor1, factor2, sep = " - ")
#     
#     # Store contrast in the list if not already present
#     contrasts[[contrast_name]] <- contrast
#   }
# }


# Construct constrast matrix
contrast_matrix2 <- makeContrasts(contrasts = contrasts, levels = design2)
colnames(contrast_matrix2) <- names(contrasts)

# Define linear model
fit2 <- eset %>%
  # linear regression of every protein against designmatrix as covariable
  lmFit(design2) %>%
  # include constrast matrix
  contrasts.fit(contrast_matrix2) %>%
  # Empirical Bayes Statistics for Differential Expression
  eBayes

```
```{r heatmap of top proteins}
top_proteins <- topTable(fit2, number = Inf)

topTable(fit2, number = Inf) %>%
  # select Protein.ID, Description and contrasts
  select(Protein.ID, Description, matches("[A-Z]_vs_B")) %>%
  # filter for proteins of interest
  filter(grepl("ester|lipase|cutinase|protease|nepenth", Description)) %>%
  # TODO add time columns, sort by time in heatmap
  # convert to long format
  pivot_longer(-c(Description, Protein.ID),
               names_to = "contrast",
               values_to = "logFC") %>%
  # plot heatmap
  ggplot(aes(x = contrast, y = Description, fill = logFC)) +
  geom_tile() +
  scale_fill_gradient2(low = "blue", mid = "white", high = "red", midpoint = 0) +
  theme_minimal() +
  # format axis labels
  theme(axis.title.x=element_blank(),
        axis.title.y=element_blank(),
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=.5, size = 6),
        axis.text.y = element_text(size = 6),
        legend.text = element_text(size = 8))


```


```{r condition:time interaction}
# Construct design matrix condition:time interaction
# does not make much sense to use interaction
# design2 <- model.matrix(~0 + CONDITION + TIME + CONDITION:TIME, data = eset)
# # create names for the condition:time interaction columns
# cond_time_interaction <- c()
# for (t in levels(eset$TIME)[-1]) {
#   for (con in levels(eset$CONDITION)[-1]) {
#     cond_time_interaction = c(cond_time_interaction, paste(con, t, sep = ":"))
#   }
# }
# # set names for the design matrix columns
# colnames(design2) <- c(levels(eset$CONDITION),
#                               levels(eset$TIME)[-1],
#                               cond_time_interaction)
```


## Analysis in DEP
```{r analysis with DEP}
library(DEP); packageVersion("DEP")

# Import fragpipe results
data <- rio::import(here::here(datadir, "combined_protein.tsv"))

# check for duplicated protein IDs and gene names
data$`Protein ID` %>% duplicated() %>% any()
data$`Gene` %>% duplicated() %>% any()
# make table of duplicated gene names
data %>% group_by(Gene) %>% summarize(frequency = n()) %>% 
  arrange(desc(frequency)) %>% filter(frequency > 1)

```


```{r remove contaminants}
# remove contaminants based on contaminant tag "cRAP..." in protein name
# where "." refers to a digit
data <- data %>%
  filter(!grepl("cRAP[0-9]{3}", Protein))

```


```{r construct SummarisedExperiment object}
data_unique <- make_unique(data, "Gene", "Protein ID", delim = ";")

# Deselect "MaxLFQ Intensities" and "Spectral Counts"
data_unique <- data_unique %>%
  select(!matches("MaxLFQ|Spectral Count"))

# Extract columns with protein intensities per sample from fragpipe output 'data'
LFQ_columns <- grep("^.*Intensity.*$", colnames(data_unique),
                    perl = TRUE)

# Remove 'Intensity' from column names
colnames(data_unique) <- colnames(data_unique) %>%
  str_replace(" Intensity", "")

# create experimental_design table for the SummarizedExperiment object
# Import sample annotation from fragpipe annotation file
sample_annotation <- rio::import(here::here(datadir, "experiment_annotation.tsv"))

# Rename condition. this will change once the proper "experiment_annotation.tsv" is used
sample_annotation <- sample_annotation %>%
  dplyr::select(sample, condition) %>%
  set_colnames(c("label", "condition")) %>%
  # using stringr package extract information from sample names
  mutate(condition = str_replace(label, "_[1-3]$", ""), # ERROR when using one letter code for condition
         replicate = str_split_i(label, "_", -1),
         time = str_split_i(label, "_", -2))
# TODO maybe use SummarizedExperimentfromExpressionSet

# Combine data to SummarisedEXperiment
data_se <- make_se(data_unique, LFQ_columns, sample_annotation)
data_se
```

```{r explore data with DEP}
# barplot of protein overlap between samples
plot_frequency(data_se)

# filter for proteins that are identified in all replicats of one condition
data_filt <- filter_missval(data_se, thr = 0)
# less stringent
data_filt2 <- filter_missval(data_se, thr = 1)

# plot no proteins per sample
plot_numbers(data_filt)

# plot protein overlap between samples
plot_coverage(data_filt)

```

```{r data normalisation with DEP}
# Normalise data
data_norm <- normalize_vsn(data_filt)

# Visualise data distribution before and after normalisation
plot_normalization(data_filt, data_norm)

```

```{r look for missing values}
# Plot a heatmap of proteins with missing values
plot_missval(data_filt)

# Plot intensity distr. and cumulative fraction of proteins with and without missing values
plot_detect(data_filt)

```
Are missing values biased towards specific samples?

Do proteins with missing values also have lo intensities?

```{r impute missing values}
# different methods are available for imputing values
# Impute missing data using random draws from a Gaussian distribution centered around a minimal value (for MNAR)
data_imp <- impute(data_norm, fun = "MinProb", q = 0.01)

# Impute missing data using random draws from a manually defined left-shifted Gaussian distribution (for MNAR)
data_imp_man <- impute(data_norm, fun = "man", shift = 1.8, scale = 0.3)

# Impute missing data using the k-nearest neighbour approach (for MAR)
#data_imp_knn <- impute(data_norm, fun = "knn", rowmax = 0.9)

# Plot intensity distributions before and after imputation
plot_imputation(data_norm, data_imp_man)
```

```{r differentiall enrichment analysis with DEP, error=TRUE}
# Differential enrichment analysis  based on linear models and empirical Bayes statistics

# Test every sample versus control
data_diff <- test_diff(data_imp, type = "control", control = "N_B_t0", design_formula = formula(~ 0 + condition + time))
```


```{r differentiall enrichment analysis with DEP 2, error=TRUE}
# Denote significant proteins based on user defined cutoffs
dep <- add_rejections(data_diff, alpha = 0.05, lfc = log2(1.5))
```
